lib/galaxy/tools/actions/__init__.py

```diff
@@ -495,9 +503,13 @@ def handle_output(name, output, hidden=None):
                                      current_job=job,
                                      out_data=out_data)
         log.info("Setup for job %s complete, ready to flush %s" % (job.log_str(), job_setup_timer))
-
         job_flush_timer = ExecutionTimer()
         trans.sa_session.flush()
+        if bulk_job_input_assocs:
+            for input_assoc in bulk_job_input_assocs:
+                input_assoc.job = job
+            trans.sa_session.bulk_save_objects(bulk_job_input_assocs)
```

Comment:
I guess we need to make sure these are happening in a single DB transaction - so the job handler doesn't pick the job up before the inputs have been written... hmm....

Reply:
Yeah - trying these tests below they fail "randomly" - making me think this is a race condition as I feared. At least it is fairly reproducible so I'll test some things.